File: src/main/java/de/gematik/test/fhir/SourceBundleBuilder.java
package de.gematik.test.fhir;

import org.hl7.fhir.r4.model.Bundle;
import org.hl7.fhir.r4.model.Parameters;
import org.hl7.fhir.r4.model.Resource;

import java.util.List;
import java.util.Objects;
import java.util.UUID;

final class SourceBundleBuilder {

  Bundle build(List<Resource> sources) {
    Objects.requireNonNull(sources, "sources must not be null");

    var bundle = new Bundle();
    bundle.setType(Bundle.BundleType.COLLECTION);

    for (var source : sources) {
      if (source == null) {
        continue;
      }
      appendResource(bundle, source);
    }

    return bundle;
  }

  private void appendResource(Bundle target, Resource source) {
    if (appendDocumentEntriesIfApplicable(target, source) || appendParameterEntriesIfApplicable(target, source)) {
      return;
    }

    addAsEntry(target, source, null);
  }

  private boolean appendDocumentEntriesIfApplicable(Bundle target, Resource source) {
    if (!(source instanceof Bundle nestedBundle) || nestedBundle.getType() != Bundle.BundleType.DOCUMENT) {
      return false;
    }

    for (var entry : nestedBundle.getEntry()) {
      if (entry.hasResource()) {
        addAsEntry(target, entry.getResource(), entry.getFullUrl());
      }
    }
    return true;
  }

  private boolean appendParameterEntriesIfApplicable(Bundle target, Resource source) {
    if (!(source instanceof Parameters parameters)) {
      return false;
    }

    for (var parameter : parameters.getParameter()) {
      for (var part : parameter.getPart()) {
        if (part.hasResource()) {
          addAsEntry(target, part.getResource(), null);
        }
      }
    }
    return true;
  }

  private void addAsEntry(Bundle target, Resource resource, String fullUrl) {
    var entry = target.addEntry();
    entry.setResource(resource);

    if (fullUrl != null && !fullUrl.isBlank()) {
      entry.setFullUrl(fullUrl);
      return;
    }

    if (resource.hasIdElement() && resource.getIdElement().hasIdPart()) {
      entry.setFullUrl("urn:uuid:" + resource.getIdElement().getIdPart());
    } else {
      entry.setFullUrl("urn:uuid:" + UUID.randomUUID());
    }
  }
}

File: src/main/kotlin/de/gematik/test/fhir/SourceBundleBuilder.kt
package de.gematik.test.fhir

import org.hl7.fhir.r4.model.Bundle
import org.hl7.fhir.r4.model.Parameters
import org.hl7.fhir.r4.model.Resource
import java.util.UUID

internal class SourceBundleBuilder {
    fun build(sources: List<Resource?>): Bundle {
        return Bundle().apply {
            type = Bundle.BundleType.COLLECTION
            sources.filterNotNull().forEach { appendResource(this, it) }
        }
    }

    private fun appendResource(target: Bundle, source: Resource) {
        when {
            appendDocumentEntriesIfApplicable(target, source) -> Unit
            appendParameterEntriesIfApplicable(target, source) -> Unit
            else -> addAsEntry(target, source, null)
        }
    }

    private fun appendDocumentEntriesIfApplicable(target: Bundle, source: Resource): Boolean {
        if (source !is Bundle || source.type != Bundle.BundleType.DOCUMENT) {
            return false
        }

        source.entry
            .filter { it.hasResource() }
            .forEach { addAsEntry(target, it.resource, it.fullUrl) }
        return true
    }

    private fun appendParameterEntriesIfApplicable(target: Bundle, source: Resource): Boolean {
        if (source !is Parameters) {
            return false
        }

        source.parameter
            .flatMap { it.part }
            .filter { it.hasResource() }
            .forEach { addAsEntry(target, it.resource, null) }
        return true
    }

    private fun addAsEntry(target: Bundle, resource: Resource, fullUrl: String?) {
        val entry = target.addEntry()
        entry.resource = resource

        if (!fullUrl.isNullOrBlank()) {
            entry.fullUrl = fullUrl
            return
        }

        entry.fullUrl = if (resource.hasIdElement() && resource.idElement.hasIdPart()) {
            "urn:uuid:${resource.idElement.idPart}"
        } else {
            "urn:uuid:${UUID.randomUUID()}"
        }
    }
}

